### PR TITLE
Refactor: Migrate bdk::FeeRate to bitcoin::FeeRate

### LINKED ISSUE #1136: Replace bdk FeeRate with rust-bitcoin FeeRate
**Describe the enhancement**  

We should be able to replace the bdk `FeeRate` struct with the `FeeRate` from rust-bitcoin since they serve the same purpose and it's less confusing to users to have only one version of the type. Any missing functions from the bdk::FeeRate should be upstreamed to rust-bitcoin or moved to a new bdk trait implemented on rust-bitcoin's `FeeRate`. 

**Use case**  

All existing uses for bdk's `FeeRate`.

**Additional context**

https://docs.rs/bitcoin/0.30.1/bitcoin/blockdata/fee_rate/struct.FeeRate.html
https://docs.rs/bdk/1.0.0-alpha.1/bdk/struct.FeeRate.html

### PR DESCRIPTION
**Pull request by junderw**
Closes #1136

### Description

This PR will eventually migrate FeeRate to rust-bitcoin's FeeRate.

### Notes to the reviewers

1. This is currently a WIP that is meant for discussion. I have made a lot of arbitrary decisions just so that we can see the full diff and go over each decision again as a team. Don't assume any decision was made with explicit intent, I just tried to make everything compile and tests pass. Everything is on the table for discussion.
2. ***If anyone wants to, feel free to take over this PR at any time.*** The legwork of figuring out all the places that need changing and which tests are affected took a few hours of trial and error. If you look at this PR and think you want to take over and finish off the more interesting parts (investigating the changes and why they were necessary / whether there are better more appropriate ways to change it) feel free. Just let me know. This issue is low priority for me personally, so don't expect this to get done in a quick manner unless someone takes it from me. I will be happy to answer questions about the new FeeRate API though.

#### Changed

- bdk::FeeRate was replaced with bitcoin::FeeRate

#### All Submissions:

* [x] I've signed all my commits
* [x] I followed the [contribution guidelines](https://github.com/bitcoindevkit/bdk/blob/master/CONTRIBUTING.md)
* [x] I ran `cargo fmt` and `cargo clippy` before committing

### CODE REVIEW

**Comment by notmandatory** [comment]
@nondiremanuel I think this PR and the associated issue #1136 deserves to be added to the alpha. We can discuss more on the next team call, but I think it needs to be done so we don't duplicate functionality with our `rust-bitcoin` dependency and if we don't do it now we won't get another chance until a BDK 2.0 release.

**Comment by notmandatory** [comment]
Thanks for taking this on @junderw ! looks good so far.

File: crates/bdk/src/wallet/coin_selection.rs [hunk 1]
```diff
@@ -276,7 +276,7 @@ impl CoinSelectionAlgorithm for OldestFirstCoinSelection {
 pub fn decide_change(remaining_amount: u64, fee_rate: FeeRate, drain_script: &Script) -> Excess {
     // drain_output_len = size(len(script_pubkey)) + len(script_pubkey) + size(output_value)
     let drain_output_len = serialize(drain_script).len() + 8usize;
-    let change_fee = fee_rate.fee_vb(drain_output_len);
+    let change_fee = (fee_rate * Weight::from_wu((drain_output_len * 4) as u64)).to_sat();
```
Comment (realeinherjar) [inline crates/bdk/src/wallet/coin_selection.rs:279]: Why are we hardcoding 4 here?
Might be worth a comment...

File: crates/bdk/tests/wallet.rs [hunk 2]
```diff
@@ -1446,7 +1446,9 @@ fn test_bump_fee_reduce_change() {
         .unwrap();
 
     let mut builder = wallet.build_fee_bump(txid).unwrap();
-    builder.fee_rate(FeeRate::from_sat_per_vb(2.5)).enable_rbf();
+    builder
+        .fee_rate(FeeRate::from_sat_per_vb_unchecked(3))
+        .enable_rbf();
```
Comment (realeinherjar) [inline crates/bdk/tests/wallet.rs:1451]: This changes the logic of the test. We are now testing an integer instead of a float.
Why?

File: crates/bdk/tests/wallet.rs [hunk 3]
```diff
@@ -1556,7 +1558,7 @@ fn test_bump_fee_reduce_single_recipient() {
     assert_eq!(tx.output.len(), 1);
     assert_eq!(tx.output[0].value + fee.unwrap_or(0), sent_received.0);
 
-    assert_fee_rate!(psbt, fee.unwrap_or(0), FeeRate::from_sat_per_vb(2.5), @add_signature);
+    assert_fee_rate!(psbt, fee.unwrap_or(0), FeeRate::from_sat_per_kwu(751), @add_signature);
```
Comment (realeinherjar) [inline crates/bdk/tests/wallet.rs:1561]: Why change from vBytes to satoshis per 1000 weight units?

File: crates/bdk/tests/wallet.rs [hunk 4]
```diff
@@ -2175,7 +2177,7 @@ fn test_fee_amount_negative_drain_val() {
     let send_to = Address::from_str("tb1ql7w62elx9ucw4pj5lgw4l028hmuw80sndtntxt")
         .unwrap()
         .assume_checked();
-    let fee_rate = FeeRate::from_sat_per_vb(2.01);
+    let fee_rate = FeeRate::from_sat_per_kwu(522);
```
Comment (realeinherjar) [inline crates/bdk/tests/wallet.rs:2180]: Why change from vBytes to satoshis per 1000 weight units?

File: crates/bdk/tests/wallet.rs [hunk 5]
```diff
@@ -2188,7 +2190,7 @@ fn test_fee_amount_negative_drain_val() {
     let psbt = builder.finish().unwrap();
     let fee = check_fee!(wallet, psbt);
 
-    assert_eq!(psbt.inputs.len(), 1);
+    assert_eq!(psbt.inputs.len(), 2);
```
Comment (realeinherjar) [inline crates/bdk/tests/wallet.rs:2193]: This seems odd to me.
From my understanding you are not changing the PSBT inputs in this test.
Hence, why this change?
Is this test broken? Was this passing before your change?

File: crates/bdk/tests/wallet.rs [hunk 6]
```diff
@@ -3369,7 +3371,7 @@ fn test_fee_rate_sign_no_grinding_high_r() {
     // alright.
     let (mut wallet, _) = get_funded_wallet("wpkh(tprv8ZgxMBicQKsPd3EupYiPRhaMooHKUHJxNsTfYuScep13go8QFfHdtkG9nRkFGb7busX4isf6X9dURGCoKgitaApQ6MupRhZMcELAxTBRJgS/*)");
     let addr = wallet.get_address(New);
-    let fee_rate = FeeRate::from_sat_per_vb(1.0);
+    let fee_rate = FeeRate::from_sat_per_kwu(251);
```
Comment (realeinherjar) [inline crates/bdk/tests/wallet.rs:3374]: Why change from vBytes to satoshis per 1000 weight units?

File: crates/bdk/tests/wallet.rs [hunk 7]
```diff
@@ -3435,7 +3437,7 @@ fn test_fee_rate_sign_grinding_low_r() {
     // signature is 70 bytes.
     let (mut wallet, _) = get_funded_wallet("wpkh(tprv8ZgxMBicQKsPd3EupYiPRhaMooHKUHJxNsTfYuScep13go8QFfHdtkG9nRkFGb7busX4isf6X9dURGCoKgitaApQ6MupRhZMcELAxTBRJgS/*)");
     let addr = wallet.get_address(New);
-    let fee_rate = FeeRate::from_sat_per_vb(1.0);
+    let fee_rate = FeeRate::from_sat_per_kwu(251);
```
Comment (realeinherjar) [inline crates/bdk/tests/wallet.rs:3440]: Why change from vBytes to satoshis per 1000 weight units?

**Comment by junderw** [comment]
To answer all of these questions:

Because FeeRate now only deals with integers, so in order to get more granularity, using sat/kwu allows us to perform <1 sat/vbyte.

As for all the other places: Because that change is what was necessary to make the tests pass.

250 sat/kwu = 1 sat/vbyte, so technically, this is also changing the value to 1.004 sat/vbyte

I have not investigated into why this change was needed to make the tests pass.

The next steps in this PR will be uncovering why each of these test changes was necessary and considering what that means for the library.

File: crates/bdk/tests/wallet.rs (hunk 5 above)
Comment (junderw) [reply crates/bdk/tests/wallet.rs:2193]: Since coin selection is affected by fees, likely the switch to a non-float FeeRate had some effect on the outcome of selected coins.

This will need investigation.

File: crates/bdk/tests/wallet.rs (hunk 2 above)
Comment (junderw) [reply crates/bdk/tests/wallet.rs:1451]: True, using kwu here to test 2.5 (which would be 625) would probably be better.

**Comment by junderw** [comment]
I edited the body of the PR:

> ***If anyone wants to, feel free to take over this PR at any time.*** The legwork of figuring out all the places that need changing and which tests are affected took a few hours of trial and error. If you look at this PR and think you want to take over and finish off the more interesting parts (investigating the changes and why they were necessary / whether there are better more appropriate ways to change it) feel free. Just let me know. This issue is low priority for me personally, so don't expect this to get done in a quick manner unless someone takes it from me. I will be happy to answer questions about the new FeeRate API though.

File: crates/bdk/src/psbt/mod.rs [hunk 8]
```diff
@@ -71,9 +71,11 @@ impl PsbtUtils for Psbt {
 
     fn fee_rate(&self) -> Option<FeeRate> {
         let fee_amount = self.fee_amount();
-        fee_amount.map(|fee| {
+        fee_amount.and_then(|fee| {
             let weight = self.clone().extract_tx().weight();
-            FeeRate::from_wu(fee, weight)
+            Some(FeeRate::from_sat_per_kwu(
```
Comment (ValuedMammal) [inline crates/bdk/src/psbt/mod.rs:76]:
```rust
fee_amount.map(|fee|  {
    Amount::from_sat(fee) / weight
})

File: crates/bdk/src/wallet/coin_selection.rs (hunk 1 above)
Comment (ValuedMammal) [reply crates/bdk/src/wallet/coin_selection.rs:279]:
```rust
let change_fee = (fee_rate * Weight::from_vb_unchecked(drain_output_len as u64)).to_sat();
```

File: crates/bdk/src/wallet/coin_selection.rs [hunk 9]
```diff
@@ -489,7 +491,7 @@ impl CoinSelectionAlgorithm for BranchAndBoundCoinSelection {
                 curr_value,
                 curr_available_value,
                 target_amount,
-                cost_of_change,
+                cost_of_change as f32,
```
Comment (ValuedMammal) [inline crates/bdk/src/wallet/coin_selection.rs:494]: can the signature of `BranchAndBoundCoinSelection::bnb` change to accept `cost_of_change: u64` as parameter ?

File: crates/bdk/src/wallet/mod.rs [hunk 10]
```diff
@@ -574,7 +574,7 @@ impl<D> Wallet<D> {
     pub fn calculate_fee_rate(&self, tx: &Transaction) -> Result<FeeRate, CalculateFeeError> {
         self.calculate_fee(tx).map(|fee| {
             let weight = tx.weight();
-            FeeRate::from_wu(fee, weight)
+            FeeRate::from_sat_per_kwu(fee.saturating_mul(1000) / weight.to_wu())
```
Comment (ValuedMammal) [inline crates/bdk/src/wallet/mod.rs:577]:
```rust
self.calculate_fee(tx)
    .map(|fee| Amount::from_sat(fee) / tx.weight())
```

File: crates/bdk/src/wallet/mod.rs [hunk 11]
```diff
@@ -1350,7 +1352,7 @@ impl<D> Wallet<D> {
             utxos: original_utxos,
             bumping_fee: Some(tx_builder::PreviousFee {
                 absolute: fee,
-                rate: fee_rate.as_sat_per_vb(),
+                rate: fee_rate.to_sat_per_vb_ceil() as f32,
             }),
```
Comment (ValuedMammal) [inline crates/bdk/src/wallet/mod.rs:1356]: this can be `rate: fee_rate` if the type of `rate` in `PreviousFee` is changed to `bitcoin::FeeRate`

File: crates/bdk/tests/psbt.rs [hunk 12]
```diff
@@ -99,8 +99,8 @@ fn test_psbt_fee_rate_with_witness_utxo() {
     assert!(finalized);
 
     let finalized_fee_rate = psbt.fee_rate().unwrap();
-    assert!(finalized_fee_rate.as_sat_per_vb() >= expected_fee_rate);
-    assert!(finalized_fee_rate.as_sat_per_vb() < unfinalized_fee_rate.as_sat_per_vb());
+    assert!(finalized_fee_rate.to_sat_per_vb_ceil() as f32 >= expected_fee_rate);
+    assert!(finalized_fee_rate.to_sat_per_vb_ceil() <= unfinalized_fee_rate.to_sat_per_vb_ceil());
 }
```
Comment (ValuedMammal) [inline crates/bdk/tests/psbt.rs:104]: If we provided a method for creating a `FeeRate` from a float with something like 
```rust
fn from_sat_per_vbf(sat_vb: f64) -> FeeRate {...} 
```

then the above can be simplified

```rust
let expected_fee_rate = FeeRate::from_sat_per_vbf(1.2345);
builder.fee_rate(expected_fee_rate);
...
assert!(finalized_fee_rate >= expected_fee_rate);
assert!(finalized_fee_rate < unfinalized_fee_rate);
```

File: crates/bdk/tests/wallet.rs [hunk 13]
```diff
@@ -135,8 +135,8 @@ fn test_get_funded_wallet_tx_fee_rate() {
     // sats are the transaction fee.
 
     // tx weight = 452 bytes, as vbytes = (452+3)/4 = 113
-    // fee rate (sats per vbyte) = fee / vbytes = 1000 / 113 = 8.8495575221 rounded to 8.849558
-    assert_eq!(tx_fee_rate.as_sat_per_vb(), 8.849558);
+    // fee rate (sats per vbyte) = fee / vbytes = 1000 / 113 = 8.8495575221 rounded up to 9
+    assert_eq!(tx_fee_rate.to_sat_per_vb_ceil(), 9);
```
Comment (ValuedMammal) [inline crates/bdk/tests/wallet.rs:139]: since we lose precision with `to_sat_per_vb_ceil`, we should assert the value in sat/kwu also.

**Comment by realeinherjar** [comment]
We don't need to implement the `floor` anymore.
`rust-bitcoin` v0.31.1 has added `to_kwu_floor` in https://github.com/rust-bitcoin/rust-bitcoin/pull/1735

**Comment by notmandatory** [comment]
@junderw can we close this PR in favor of #1216 ?

**Comment by junderw** [comment]
Of course!
